Content.Shared/Temperature/Systems/SharedTemperatureSystem.cs

```diff
@@ -119,10 +142,83 @@ public float GetHeatCapacity(Entity<TemperatureComponent?> entity)
     {
         var (uid, temperature) = entity;
 
-        if (!TemperatureQuery.Resolve(uid, ref temperature) || !_physicsQuery.TryGetComponent(uid, out var physics) || physics.FixturesMass <= 0)
+        if (!TemperatureQuery.Resolve(uid, ref temperature))
             return Atmospherics.MinimumHeatCapacity;
 
-        return temperature.SpecificHeat * physics.FixturesMass;
+        if (temperature.HeatCapacityDirty)
+            UpdateCachedHeatCapacity((uid, temperature));
+
+        return temperature.CachedHeatCapacity;
+    }
+
+    /// <summary>
+    /// Recalculates the heat capacity of an entity as necessary.
+    /// </summary>
+    /// <param name="entity">The entity that needs to have its heat capacity updated.</param>
+    private void UpdateCachedHeatCapacity(Entity<TemperatureComponent> entity)
+    {
+        var (uid, temperature) = entity;
+
+        DebugTools.Assert(temperature.HeatCapacityDirty, $"Recalculated the heat capacity of {ToPrettyString(entity)} without it having been dirtied.");
+        temperature.HeatCapacityDirty = false;
+
+        var ev = new HeatCapacityUpdateEvent(entity, temperature.BaseHeatCapacity);
+        RaiseLocalEvent(uid, ref ev);
+        temperature.CachedHeatCapacity = ev.HeatCapacity;
+
+        DebugTools.Assert(!temperature.HeatCapacityDirty, $"The heat capacity of {ToPrettyString(entity)} was dirtied while it was being recalculated.");
+    }
+
+    /// <summary>
+    /// Dirties the heat capacity of an entity.
+    /// </summary>
+    /// <param name="entity">The entity to dirty the heat capacity of.</param>
+    public void DirtyHeatCapacity(Entity<TemperatureComponent?> entity)
+    {
+        if (!TemperatureQuery.Resolve(entity, ref entity.Comp))
+            return;
+
+        entity.Comp.HeatCapacityDirty = true;
+    }
+
+    /// <summary>
+    /// Sets the base heat capacity of an entity.
+    /// </summary>
+    /// <remarks>
+    /// Does not affect the heat capacity contributed by the mass of the entity or other factors.
+    /// </remarks>
+    /// <param name="entity">The entity to change the base heat capacity of.</param>
+    /// <param name="value">The new base heat capacity for the entity.</param>
+    public void SetHeatCapacity(Entity<TemperatureComponent?> entity, float value)
```

Comment:
This method could have energy conservation issues, e.g. having a TemperatureComponent at some temperature $T$, then setting the heat capacity causes $T$ to stay the same, resulting in energy conservation issues.

Reply:
This is basically just for VV use and if an admin is messing with that I don't think they want them messing with heat capacity causing the temperature to bounce around the place.